### PR TITLE
php: expose mkExtension

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
- make it easier to build extensions using `nix-phps`, as mentioned in https://github.com/fossar/nix-phps/issues/27#issuecomment-869742250

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by etu** [comment]
I'm thinking we should document this with example usage as well, including
how to use a custom built extension.

This is a good thing to make it easier to patch or customize extensions and
such.

**Comment by aanderse** [comment]
@etu sounds good. Do you have an example in mind? I think we package all the included extensions, so a nice "clean" example doesn't come to mind. This is my intended use case:
```
mysql = lib.mkIf (lib.versionOlder prev.php.version "7.0") (prev.mkExtension {
  name = "mysql";
  internalDeps = [ prev.php.extensions.mysqlnd ];
  configureFlags = [ "--with-mysql" "--with-mysql-sock=/run/mysqld/mysqld.sock" ];
});
```

**Comment by jtojnar** [comment]
Yeah, this is mostly something internal for our convenience, rather than something anyone would be expected to use so I would not really promote it in docs.

File: pkgs/top-level/php-packages.nix [hunk 1]
```diff
@@ -59,6 +59,72 @@ lib.makeScope pkgs.newScope (self: with self; {
     pname = "php-${pname}";
   });
 
+  # Function to build a single php extension based on the php version.
```
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:62]: Maybe mention that this function is internal and that it builds extensions shipped as part of the PHP project.

We might even call it `mkBundledExtension` or something.

Comment (aanderse) [reply pkgs/top-level/php-packages.nix]: I updated the documentation to this:
```
  # Function to build an extension which is shipped as part of the php
  # source, based on the php version.
```

Personally, I think `mkBundledExtension` is a bit of a mouthful, so I left it as `mkExtension`. I don't care too much, though, so if anyone prefers `mkBundledExtension` let me know and I'll rename it.

**Comment by aanderse** [comment]
@etu @jtojnar good to merge?

**Comment by aanderse** [comment]
Thanks team!

**Comment by aanderse** [comment]
@etu & @talyz should we consider adding a `postPatch` argument to `mkExtension` to avoid things like this:

https://github.com/NixOS/nixpkgs/blob/f091739337022031ebeb14a8f1bc2ecfc76d14fc/pkgs/top-level/php-packages.nix#L269-L273

It seems like it could be helpful in general. If you like I could write a PR to do so.

**Pull request merged**